File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:

        res = []
        curr = []

        def dfs(i):
            #out-of-bounds
            if i >= len(nums):
                res.append(curr.copy())
                return curr


            #include one that adds
            curr.append(nums[i])
            dfs(i + 1)

            #includ oen that pops
            curr.pop()
            dfs(i + 1)
        
        dfs(0)

        return res
        
        

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) 
    {
        std::vector<int> curr;

        dfs(nums, curr, 0);

        return res;
    }

    void dfs(vector<int>& nums, vector<int> curr, int i)
    {

        //Base Case
        if (i >= nums.size())
        {
            res.push_back(curr);
            return;
        }

        //Add
        curr.push_back(nums[i]);

        //Take 
        dfs(nums, curr, i + 1);

        //Skip
        curr.pop_back();
        dfs(nums, curr, i + 1);
        
    }

private:
    vector<vector<int>> res;
};